csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -1384,8 +1453,19 @@ private module PostUpdateNodes {
 private import PostUpdateNodes
 
 /** A node that performs a type cast. */
-class CastNode extends ExprNode {
-  CastNode() { this.getExpr() instanceof CastExpr }
+class CastNode extends Node {
+  CastNode() {
+    this.asExpr() instanceof Cast
+    or
+    exists(Ssa::ExplicitDefinition def |
```

Comment:
The name `CastNode` is the one used in the interface of the shared data flow implementation, and really it means any node where we may be able to stop flow based on the type of the tracked object.

The [first case](https://github.com/Semmle/ql/pull/2308/files/54088248a17760c30791a6e1a581211d4d2adb2e..abcb6b8aab969a58df3c7cb2b1ca53d8f398bc2c#diff-623b51985c523c1ec3d03ed19b4ef882R1460) adds a type check for patterns. For example, if the object we are tracking is of type `int`, then there will be no flow to the SSA definition for `s` in `x is string s`.

The [second and third cases](https://github.com/Semmle/ql/pull/2308/files/54088248a17760c30791a6e1a581211d4d2adb2e..abcb6b8aab969a58df3c7cb2b1ca53d8f398bc2c#diff-623b51985c523c1ec3d03ed19b4ef882R1465-R1467) add similar type checks, but this time for reads/writes to fields. The reason we need that is because fields may have a generic type, so we may be able to stop flow by checking the type, as in [this](https://github.com/Semmle/ql/pull/2308/files#diff-f1d2bf86863efbda7e15d3476d50e447R100) example.

Reply:
An alternative to adding the second and third cases, is to refine [`TFieldLikeContent`](https://github.com/Semmle/ql/pull/2308/files#diff-623b51985c523c1ec3d03ed19b4ef882R443) to track also the (GVN) type (i.e. more or less reverting the addition of `f.getSourceDeclaration() = f `). We then need to refine `readStep` and `storeStep` instead, basically moving the `compatiableTypes` check there. I may end up making that change, but rather in a follow-up PR.